loom/common/core/telemetry.cpp

```diff
@@ -42,133 +48,277 @@ void Telemetry::disable()
 void Telemetry::beginTick()
 {
     if (enabled != pendingEnabled) {
+        if (pendingEnabled) tickThreadId = platform_getCurrentThreadId();
         enabled = pendingEnabled;
         lmLog(gTelemetryLogGroup, "Telemetry %s", enabled ? "enabled" : "disabled");
     }
     if (!enabled) return;
 
     tickValues.reset();
-    tickRanges.reset();
-    tickTimerStack.clear();
+    tickProfilerRootsVisited = 0;
 
     loom_resetTimer(tickTimer);
 
+    // Customized asset protocol message (3 ints + telemetry + other tables)
+    sendBuffer.setPosition(0);
+
+    // Size of the whole message, late write
+    sendBuffer.writeUnsignedInt(-1);
+    sendBuffer.writeInt(0xDEADBEEF);
+    sendBuffer.writeInt(LOOM_FOURCC('T', 'E', 'L', 'E'));
+
+    // Size of telemetry events, late write
+    sendBuffer.writeUnsignedInt(-1); 
+
+    // Endianness of the device for more efficient recording
+    sendBuffer.writeUnsignedInt(UT_ENDIAN);
+
+    // Event stream starting point for later use
+    eventsStartPos = sendBuffer.getPosition();
+
     // Add the tick id as a default tick value
     setTickValue("tick.id", tickId);
+
+    tickProfilerActive = true;
 }
 
-void Telemetry::endTick()
+// Used to build up a telemetry root structure while reading.
+struct TickProfilerRoot {
+    utString name;
+    int count;
+};
+
+bool Telemetry::readTickProfiler(utByteArray& buffer, JSON& ranges, JSON& meta)
 {
-    if (!enabled) return;
+    loom_resetTimer(tickTimer);
 
-    lmAssert(tickTimerStack.size() == 0, "Tick timer end call missing");
+    size_t eventsSize = buffer.readUnsignedInt();
+    unsigned int eventEndianness = buffer.readUnsignedInt();
+    size_t eventsPos = buffer.getPosition();
+    
+    // Skip to the profiler root tree
+    buffer.setPosition(eventsPos + eventsSize);
+
+    utHashTable<utUInt64HashKey, TickProfilerRoot> addrToName;
```

Comment:
Reserving this would be a good idea, yes. Maybe we should add a size at the beginning so we can preallocate completely. We can reserve ancestors, but I don't think it'll have much of an effect as it's only as big as the tree depth.

Both of these are eclipsed by JSON allocation / processing / generation though.